README.md
# Here Unicode Mate
This extension is the best tool helps you work with Unicode and CJK encodings in Visual Studio Code.

本外掛是 VSCode 上處理 Unicode 與 CJK 編碼最方便的工具。

このプラグインは VSCode でユニコードや CJK 文字コードを扱うのに一番便利なツールです。

# Functions | 功能 | 機能

## Show Unicode value | 顯示字元 Unicode 值 | 文字コード表示

This extension shows Unicode value for the current character of keyboard cursor in the status bar.

隨時將目前游標處字元的 Unicode 值顯示在狀態列上。

現在カーソル位置の文字の Unicode の文字コードがステータスバーに表示されます。

![image](https://user-images.githubusercontent.com/5418570/82236077-2c780b00-9966-11ea-83c4-e335b0e92141.png)

If the character comes with combining marks, variation selectors or zero-width-joiner (e.g. emojis), the whole sequence will be shown.

當字元後方有組合字元、異體字選擇子、連接字元（如繪文字等）時，能夠顯示出完整序列。

文字の後ろに結合文字・異体字セレクタまたはジョイナー（複雑な絵文字など）がついている場合、まとめて表示されます。

![image](https://user-images.githubusercontent.com/5418570/82236168-4d406080-9966-11ea-9a5a-7b2265a7d702.png)

## Encoding forms | 各種編碼方式 | 符号化方式

When you click the Unicode value in the status bar, many kinds of encoding forms will be shown.
Such as UTF-8, UTF-16 (Surrogate Pairs), entity refernce of HTML and URL encode.

點選狀態列上的 Unicode 值，能夠顯示出此字元的各種編碼方式。像是 UTF-8、UTF-16（代理組）與 HTML 實體參照等，不用再自己計算了。

ステータスバーに表示される Unicode 値をクリックすると、様々な符号化方式が表示されます。UTF-8、UTF-16サロゲートペア、HTML 実体参照などの複雑な計算はもうしなくて済みます。

![image](https://user-images.githubusercontent.com/5418570/82236450-b1632480-9966-11ea-9747-4341cd08bf98.png)

If the character is contained in CJK encodings or Adobe ROS, the ANSI code or CID code will also be shown.
Big5, Big5-HKSCS, Big-UAO, Shift-JIS, GBK and KS are supported.

若此字元包含於各種 CJK 編碼或 Adobe ROS，也會一並顯示出字碼或 CID 值。
支援 Big5、Big5 香港增補字符集、Big5-UAO (Unicode補完計畫)、Shift-JIS、GBK 與 KS 等。

CJK 文字の固有符号化方式の文字コードと Adobe ROS の CID コードにも対応。Windows と Mac の違いや、JIS83 と JIS2004 の違いも対応。

![image](https://user-images.githubusercontent.com/5418570/82473452-287df180-9afc-11ea-8d62-f9f659768068.png)

We also support the glyph Nice Names of [Glyphs](https://glyphsapp.com/)!

並支援 [Glyphs](https://glyphsapp.com/) 的預設字符名稱。

[Glyphs](https://glyphsapp.com/) のグリフのナイスネームも表示可能！


You can select any line to copy it, and paste to your source code!

您可選取任何一行將字串複製到剪貼簿，貼到程式碼或應用程式裡使用。

表示される文字列を選択し、クリップボードにコピーされます。自由にソースコードなどに貼り付けてご利用ください。

## Reverse pattern | 反解也可以 | 逆パターン

We can know the Unicode (and many other encodings) of character now.
However, can we know what the code is if we have a Unicode value?

Don't worry. You can just move the mouse cursor over the Unicode value. And the character information will be popuped. Easy and simple!

有時候反而是會在程式碼裡碰到已經編碼過的 Unicode 值，但看不出來是什麼字元，要換算也很麻煩。
不過沒關係，本工具讓您只要將滑鼠游標移到上面，就直接會顯示出這是什麼字元。簡單吧！

逆パターンもOK。ソースコードなどですでにエンコードされた文字コードを目にすることがあります。なんの文字かわかりにくい場合も多いはず（特にUTF-8・サロゲートなどは計算しにくい）。
大丈夫！マウスカーソルを指すだけで、文字情報が表示されます。

![image](https://user-images.githubusercontent.com/5418570/82473760-a2ae7600-9afc-11ea-852d-bb86396939b6.png)

## Unicode character input helper | 輸入幫手 | 入力ヘルパー

Just press Ctrl + 🄸 (or Cmd⌘ + 🄸 in Mac), unicode name list will be shown.
You can input some keyword to query the character which you want to input.

按 Ctrl + 🄸 (或 Cmd⌘ + 🄸)，會顯示出所有 Unicode 字元名稱。
在這裡可以輸入一些簡單的關鍵字，輔助輸入不容易輸入的字元。

Ctrl + 🄸 (または Cmd⌘ + 🄸) を押すと、全ての文字の Unicode 名などが表示される。
簡単なキーワードで入力したい文字を探し出して入力してください。

![image](https://user-images.githubusercontent.com/5418570/82473976-eacd9880-9afc-11ea-9262-080cc2b3276b.png)

![image](https://user-images.githubusercontent.com/5418570/82474277-662f4a00-9afd-11ea-876e-40dad6c83cc9.png)

This is very helpful to input special symbols, emojis, and foreign language letters which you don't have their IMEs.

本功能適合用來輸入特殊符號、繪文字，甚至可以作為簡易的輸入法輸入外文字母。

この機能は記号・絵文字の入力に活躍します。また、IMEのインストールが面倒な場合、外国語の文字の入力にも使えます。


Hint / 使用例提示 / 入力ヒント ：

- ¢ : cent
- ฿ : baht
- 한 : han
- ห : hoh
- 😍 : smil
- 🕒 : threeoc
- 🀄 : red
- … : ...
- 🈲 : 禁

# To be done

- to implement complete [Grapheme Cluster Boundaries](http://www.unicode.org/reports/tr29/#Grapheme_Cluster_Boundaries) (UAX#29) to support more language such as Arabic, Indian, ...
- to support variation selectors of Adobe ROSs.
- ...


CHANGELOG.md
# Change Log
All notable changes to the "cursorCharCode" extension will be documented in this file.

### v0.9.2 (2020/5/23)

- Enhance Unicode character input helper.

### v0.9.1 (2020/5/21)

- Rename the extension.
- Unicode character input helper. (just try Ctrl+I)
- Fix UTF-8 bugs.
- Support Python UTF-32 style in hover hint mode.

### v0.9.0 (2020/5/18)

- Beta release.


makedb/ucd2json.rb
require 'json'

codes = {}
ranges = {}

f = File.open('./UnicodeData.txt', 'r')
f.each { |s|
	s.chomp!
	tmp = s.split(';')
	u = tmp[0]
	n = tmp[1]

	# F900;CJK COMPATIBILITY IDEOGRAPH-F900;Lo;0;L;8C48;;;;N;;;;;
	if n =~ /CJK COMPATIBILITY/
		n.gsub!(/-([0-9A-F]+)/, ' (' + tmp[5] + ')')
	end

	n.gsub!(/\w+/, &:capitalize).gsub!(/Cjk/, 'CJK')

	if n[0] != '<'
		codes[u] = n #.capitalize
	elsif tmp[10] != nil && tmp[10] != ""
		codes[u] = tmp[10].gsub!(/\w+/, &:capitalize)
	elsif n =~ /<(.+), First>/
		r = $1 #.capitalize
		ranges[r] = [u, nil]
	elsif n =~ /<(.+), Last>/
		r = $1 #.capitalize
		ranges[r][1] = u
	end
}
f.close

f = File.open('./ucd.json', 'w:utf-8')
f.puts JSON.pretty_generate({'codes' => codes, 'ranges' => ranges})
#f.puts JSON.generate({'codes' => codes, 'ranges' => ranges}).
f.close



src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the necessary extensibility types to use in your code below
import {window, env, Disposable, ExtensionContext, StatusBarAlignment, StatusBarItem, Uri, Range, commands, TextEditor, languages, Hover} from 'vscode';
import * as path from 'path';
import { totalmem } from 'os';

//let ucd, enc;

function encodeDec(decs: number[], func: (number) => string) {
    let res = '';
    for (var i in decs) res += func(decs[i]);
    return res;
}

function encodeHex(hexs: string[], func: (string) => string) {
    let res = '';
    for (var i in hexs) res += func(hexs[i]);
    return res;
}

function toUTF8(code: number) {
    if (code <= 0x7f) return '\\x' + toHex(code, 2);
    if (code <= 0x7ff) return '\\x' + toHex((code >> 6) | 0xc0, 2) + '\\x' + toHex((code & 0x3f) | 0x80, 2);
    if (code <= 0xffff) return '\\x' + toHex((code >> 12) | 0xe0, 2) + 
        '\\x' + toHex(((code >> 6) & 0x3f) | 0x80, 2) + '\\x' + toHex((code & 0x3f) | 0x80, 2);
    return '\\x' + toHex((code >> 18) | 0xf0, 2) + '\\x' + toHex(((code >> 12) & 0x3f) | 0x80, 2) +
        '\\x' + toHex(((code >> 6) & 0x3f) | 0x80, 2) + '\\x' + toHex((code & 0x3f) | 0x80, 2);

    //if (code >= 0x10000) return '\\x' + toHex((n & 0x07) | 0xf0, 2) + res;
    //if (code >= 0x800) return '\\x' + toHex((n & 0x0f) | 0xe0, 2) + res;
    //return '\\x' + toHex((n & 0x1f) | 0xc0, 2) + res;
}

const cidRangeData = {
    'aj': [8283, 8358, 8719, 9353, 15443, 20316, 23057, 23059],
    'ac': [14098, 17407, 17600, 18845, 18964, 19087, 19155, 19178],
    'ag': [7716, 9896, 22126, 22352, 29063, 30283],
    'aks': [9332, 18154, 18351],
    'ak': [3058, 4636, 11450, 11730, 11877, 12234, 14237, 18857, 22479, 22896]
};

function cidRange(ro, cid) {
    for (var i=0; i<cidRangeData[ro].length; i++) {
        if (cid <= cidRangeData[ro][i]) return i+'';
    }
}


// This method is called when your extension is activated. Activation is
// controlled by the activation events defined in package.json.
export function activate(context: ExtensionContext) {
    let charCodeDisplay = new CharCodeDisplay();
    let controller = new CharCodeController(charCodeDisplay);

    const ucd = require(context.asAbsolutePath(path.join('resource', 'ucd.json')));
    const enc = require(context.asAbsolutePath(path.join('resource', 'encodings.json')));

    let makeCodeHover = function(title: string, u: string, type: string) {
        u = u.toUpperCase();
        let c = String.fromCodePoint(parseInt(u, 16));
        let n = null;
        if (ucd.codes[u]) {
            n = ucd.codes[u];
        } else {
            for (var i in ucd.ranges) {
                if (u >= ucd.ranges[i].start && u <= ucd.ranges[i].end) n = ucd.ranges[i].name;
                //if (u >= ucd.ranges[rn][0] && u <= ucd.ranges[rn][1]) n = rn;
            }
        }
        return new Hover(`${title} **${c}**\n\nU+${u} in *${type}*\n\n**${n}**`);
    };

    function toNameItem(c:string, str:string, note:string) {
        return {label: str + ' (' + String.fromCodePoint(parseInt(c, 16)) + ')', description: 'U+' + c + (note != null ? ' ' + note : '')};
    }

    let createCharNameList = function() {
        var res = [];
        for (var c in ucd.codes) res.push(toNameItem(c, ucd.codes[c], null));
        for (var c in enc) {
            if (enc[c].nk) res.push(toNameItem(c, enc[c].nk, null));
            if (enc[c].sk) res.push(toNameItem(c, enc[c].sk, ucd.codes[c] || enc[c].nn || enc[c].nk || null));
        }
        return res;
        /*
        var res = [];
        var tmp = {};
        for (var c in ucd.codes) tmp[c] = ucd.codes[c];
        for (var c in enc) {
            if (! enc[c].nn) continue;
            if (tmp[c]) tmp[c] += ' / ' + enc[c].nn; else tmp[c] = enc[c].nn;
        }
        
        for (var c in tmp) if (c.length == 4)
            res.push({label: tmp[c] + ' (' + String.fromCodePoint(parseInt(c, 16)) + ')', description: 'U+' + c});
        for (var c in tmp) if (c.length == 5)
            res.push({label: tmp[c] + ' (' + String.fromCodePoint(parseInt(c, 16)) + ')', description: 'U+' + c});
        return res;
        */
    };

    const charNames = createCharNameList();
    //console.log(charNames.length);

    // Add to a list of disposables which are disposed when this extension is deactivated.
    context.subscriptions.push(controller);
    context.subscriptions.push(charCodeDisplay);

    context.subscriptions.push(
        commands.registerCommand('cursorCharcode.detailUnicodeInfo', async () => {
            if (charCodeDisplay.unicodeDecList.length == 0) return;


            let chars = charCodeDisplay.characters;
            let decs = charCodeDisplay.unicodeDecList;
            let hexs = charCodeDisplay.unicodeHexList;

            //surrogate pairs
            let sdecs = [];
            let shexs = [];
            for (let i=0; i < chars.length; i++) {
                let code = chars.charCodeAt(i);
                sdecs.push(code);
                shexs.push(toHex(code, 4));
            }
            
            let list = [
                { label: encodeHex(hexs, (u) => 'U+' + u + ' '), description: 'Unicode' },
                { label: encodeDec(decs, (u) => toUTF8(u)), description: 'UTF-8' },
                { label: encodeHex(shexs, (u) => '\\u' + u), description: 'UTF-16' },
                { label: encodeDec(decs, (u) => '&#' + u + ';'), description: 'HTML Entity Reference' },
                //{ label: encodeHex(hexs, (u) => '&#x' + u + ';'), description: 'Entity Reference' },
                { label: encodeURIComponent(chars), description: 'URL Encode' }
            ];

            let nameList = encodeHex(hexs, (u) => {
                if (ucd.codes[u]) {
                    return ' / ' + ucd.codes[u];
                } else {
                    for (var i in ucd.ranges) {
                        if (u >= ucd.ranges[i].start && u <= ucd.ranges[i].end) return ' / ' + ucd.ranges[i].name;
                        //if (u >= ucd.ranges[rn][0] && u <= ucd.ranges[rn][1]) return ' / ' + rn;
                    }
                }
            });
            list.push({ label: nameList.substring(3), description: 'Unicode Name' });

            var h = hexs[0];
            if (enc[h]) {
                if (enc[h].b5) list.push({ label: '0x' + enc[h].b5, description: 'Big5' });
                if (enc[h].mb) list.push({ label: '0x' + enc[h].mb, description: 'Big5 (Mac)' });
                if (enc[h].hk) list.push({ label: '0x' + enc[h].hk, description: 'Big5-HKSCS' });
                if (enc[h].ua) list.push({ label: '0x' + enc[h].ua, description: 'Big5-UAO' });
                if (enc[h].sj) list.push({ label: '0x' + enc[h].sj, description: 'ShiftJIS' });
                if (enc[h].mj) list.push({ label: '0x' + enc[h].mj, description: 'ShiftJIS (Mac)' });
                if (enc[h].gb) list.push({ label: '0x' + enc[h].gb, description: 'GBK' });
                if (enc[h].ks) list.push({ label: '0x' + enc[h].ks, description: 'KS (Wansung)' });
                if (enc[h].aj) list.push({ label: '/' + enc[h].aj, description: 'Adobe-Japan1-' + cidRange('aj', enc[h].aj) });
                if (enc[h].ajn) list.push({ label: '/' + enc[h].ajn, description: 'Adobe-Japan1-' + cidRange('aj', enc[h].ajn) + 'N' });
                if (enc[h].ac) list.push({ label: '/' + enc[h].ac, description: 'Adobe-CNS1-' + cidRange('ac', enc[h].ac) });
                if (enc[h].ag) list.push({ label: '/' + enc[h].ag, description: 'Adobe-GB1-' + cidRange('ag', enc[h].ag) });
                if (enc[h].ak) list.push({ label: '/' + enc[h].ak, description: 'Adobe-KR-' + cidRange('ak', enc[h].ak) });
                if (enc[h].aks) list.push({ label: '/' + enc[h].aks, description: 'Adobe-Korea1-' + cidRange('aks', enc[h].aks) });
            }

            let glist = encodeHex(hexs, (u) => {
                if (enc[u] && enc[u].nn) return ', ' + enc[u].nn;
                if (enc[u] && enc[u].nk) return ', ' + enc[u].nk;
                if (u.length == 4 && u >= 'FE00' && u <= 'FE0f') return '';     // Variation Selectors (SVS)
                if (u.length == 5 &&u >= 'E0100' && u <= 'E01EF') return '';    // Variation Selectors Supplement (IVS)
                return u.length == 4 ? ', uni' + u : ', u' + u;
            });
            if (glist.length > 0) list.push({ label: glist.substring(2), description: 'Glyphs Nice Name' });
            //for (var i in hexs) {
            //    list.push({ label: "More about U+" + hexs[i] + "...", description: 'Open Browser' });
            //}

            //window.showInformationMessage(charCodeDisplay.hexCode);
            let val = await window.showQuickPick(list, {placeHolder: 'Select to copy'});
            if (val) {
                if (val.description == 'Glyphs Nice Name') {
                    env.clipboard.writeText(val.label.replace(/, /g, "\n"));
                //} else if (val.description == 'Open Browser') {
                //    if (val.label.match(/U\+([0-9A-F]+)/)) 
                //        commands.executeCommand('avscode.open', Uri.parse('https://unicode-table.com/en/' + RegExp.$1));
                } else {
                    env.clipboard.writeText(val.label);
                }
            }
        })
    );

    context.subscriptions.push(
        commands.registerCommand('cursorCharcode.queryUnicodeChar', async () => {
            let val = await window.showQuickPick(charNames, {placeHolder: 'Input keywords to search character'});
            if (val) {
                var editor = window.activeTextEditor;
                editor.edit(function(edit) {
                    edit.delete(editor.selection);
                }).then(function() {
                    editor.edit(function(edit) {
                        edit.insert(editor.selection.start, String.fromCodePoint(parseInt(val.description.substring(2), 16)));
                    });
                });
            }
        })
    );

    context.subscriptions.push(
        languages.registerHoverProvider('*', {
            provideHover(document, position, token) {
                let r = document.getWordRangeAtPosition(position, /&#([0-9A-Fa-fx]+);/i);
                let w = null;
                if (r) {
                    w = document.getText(r);
                    if (w.match(/&#(\d+);/)) {
                        let u = parseInt(RegExp.$1, 10);
                        return makeCodeHover(`&amp;#${u};`, toHex(u, 4), 'HTML Entity Reference (Decimal)');
                    } else if (w.match(/&#x([0-9A-F]+);/)) {
                        let u = RegExp.$1;
                        return makeCodeHover(`&amp;#x${u};`, u, 'HTML Entity Reference (HEX)');
                    } 
                }

                r = document.getWordRangeAtPosition(position, /U\+([0-9A-Fa-f]{4,5})/);
                if (r) {
                    //w = document.getText(r);
                    //if (w.match(/U\+([0-9A-F]{4,5})/i)) {
                    let u = RegExp.$1;
                    return makeCodeHover('U+'+u, u, 'Unicode');
                    //}
                }

                r = document.getWordRangeAtPosition(position, /\\x([2-7][0-9A-Fa-f])/);
                if (r) {
                    //w = document.getText(r);
                    let u = RegExp.$1;
                    return makeCodeHover(`\\x${u}`, '00'+u, 'ASCII');
                }
                
                r = document.getWordRangeAtPosition(position, /\\x[C-Fc-f][0-9A-Fa-f](\\x[89ABab][0-9A-Fa-f])+/);
                if (r) {
                    w = document.getText(r);
                    if (w.match(/\\x(F[0-7])\\x([89AB][0-9A-F])\\x([89AB][0-9A-F])\\x([89AB][0-9A-F])/i)) { // UTF-8 (4byte)
                        let b1 = RegExp.$1;
                        let b2 = RegExp.$2;
                        let b3 = RegExp.$3;
                        let b4 = RegExp.$4;
                        let u = ((parseInt(b1, 16) & 0x07) << 18) | ((parseInt(b2, 16) & 0x3f) << 12) | ((parseInt(b3, 16) & 0x3f) << 6) | (parseInt(b4, 16) & 0x3f);
                        return makeCodeHover(`\\x${b1}\\x${b2}\\x${b3}\\x${b4}`, u.toString(16), 'UTF-8');
                    } else if (w.match(/\\x([EF][0-9A-F])\\x([89AB][0-9A-F])\\x([89AB][0-9A-F])/i)) { // UTF-8 (3byte)
                        let b1 = RegExp.$1;
                        let b2 = RegExp.$2;
                        let b3 = RegExp.$3;
                        let u = ((parseInt(b1, 16) & 0x0f) << 12) | ((parseInt(b2, 16) & 0x3f) << 6) | (parseInt(b3, 16) & 0x3f);
                        return makeCodeHover(`\\x${b1}\\x${b2}\\x${b3}`, toHex(u, 4), 'UTF-8');
                    } else if (w.match(/\\x([CD][0-9A-F])\\x([89AB][0-9A-F])/i)) { // UTF-8 (2byte)
                        let b1 = RegExp.$1;
                        let b2 = RegExp.$2;
                        let u = ((parseInt(b1, 16) & 0x1f) << 6) | (parseInt(b2, 16) & 0x3f);
                        return makeCodeHover(`\\x${b1}\\x${b2}`, toHex(u, 4), 'UTF-8');
                    } 
                }

                r = document.getWordRangeAtPosition(position, /\\u([Dd][89ABab][0-9A-Fa-f]{2})\\u([Dd][C-Fa-f][0-9A-Fa-f]{2})/);
                if (r) {
                    //w = document.getText(r);
                    let b1 = RegExp.$1;
                    let b2 = RegExp.$2;
                    let u = (((parseInt(b1, 16) & 0x3ff) << 10)) + (parseInt(b2, 16) & 0x3ff) + 0x10000;
                    return makeCodeHover(`\\u${b1}\\u${b2}`, u.toString(16), 'UTF-16 Surrogate Pairs');
                }

                r = document.getWordRangeAtPosition(position, /\\u([0-9A-Fa-f]{4})/);
                if (r) {
                    //w = document.getText(r);
                    let u = RegExp.$1;
                    return makeCodeHover(`\\u${u}`, toHex(parseInt(u, 16), 4), 'UTF-16 BMP');
                }

                r = document.getWordRangeAtPosition(position, /\\U([0-9A-Fa-f]{8})/);
                if (r) {
                    //w = document.getText(r);
                    let u = RegExp.$1;
                    return makeCodeHover(`\\U${u}`, u, 'UTF-32');
                }
                
                return null;
            }
        })
    );
}

function toHex(num: number, padLength: number) {
    let s = num.toString(16).toUpperCase();
    return s.length >= padLength ? s : '0'.repeat(padLength - s.length) + s;
}

function checkSubchar(c: number) {

    // To be done
    // To implement by Grapheme Cluster Boundaries (UAX#29)
    // http://www.unicode.org/reports/tr29/#Grapheme_Cluster_Boundaries
    // Mn  https://www.compart.com/en/unicode/category/Mn

    if (c >=  0x0300 && c <=  0x036f) return true;        // Combining character
    if (c >=  0xfe00 && c <=  0xfe0f) return true;        // Variation Selectors (SVS)
    if (c >= 0x1f3fb && c <= 0x1f3ff) return true;        // Emoji Modifier Fitzpatrick
    if (c >= 0xe0100 && c <= 0xe01ef) return true;        // Variation Selectors Supplement (IVS)

    if (c >=  0x0483 && c <=  0x0487) return true;        // Combining Cyrillic
    if (c >=  0x0591 && c <=  0x05c7) return true;        // Combining Hebrew

    if (c ==  0x0e31)                 return true;        // Thai
    if (c >=  0x0e33 && c <=  0x0e3a) return true;        // Thai
    if (c >=  0x0e47 && c <=  0x0e4e) return true;        // Thai

    if (c >=  0x180b && c <=  0x180d) return true;        // Mongolian Variation Selectors (FVS)
    if (c >=  0x1ab0 && c <=  0x1aff) return true;        // Combining Diacritical Marks Extended
    if (c >=  0x1dc0 && c <=  0x1dff) return true;        // Combining Diacritical Marks Supplement
    if (c >=  0x20d0 && c <=  0x20ff) return true;        // Combining Diacritical Marks for Symbols
    if (c >=  0xfe20 && c <=  0xfe2f) return true;        // Combining Half Marks
    
    return false;
}

class CharCodeDisplay {
    private _statusBarItem: StatusBarItem;
    //private _charRange: Range;
    //private _character: string;
    //private _hexCode: string;
    private _characters: string;
    private _unicodeDecList: number[];
    private _unicodeHexList: string[];

    /**
     * Returns the range of the character in the active editor.
     */
    //public get charRange() { return this._charRange; }

    public get characters() { return this._characters; }
    public get unicodeDecList() { return this._unicodeDecList; }
    public get unicodeHexList() { return this._unicodeHexList; }

    public updateCharacterCode(editor?: TextEditor) {
        if(!this._statusBarItem) {
            this._statusBarItem = window.createStatusBarItem(StatusBarAlignment.Left);
        }

        // Get the current text editor
        if(!editor) {
            editor = window.activeTextEditor;
        }

        if(!editor || !editor.selection || !editor.document) {
            this._statusBarItem.hide();
            return;
        }

        let cursorPos = editor.selection.active;
        let pos = cursorPos;

        let charDecList = [];
        let tryNext = true;
        let joinerOn = false;

        do {
            // taking 2 chars instead of one allows to handle surrogate pairs correctly
            let textRange = new Range(pos, pos.translate(0, 2));
            let text = editor.document.getText(textRange);
            if (!text) break;
            let charDec = text.codePointAt(0);
            if (!charDecList) break;

            if (charDecList.length > 0) {
                if (charDec == 0x200d) {        // // zero width joiner
                    joinerOn = true;
                } else if (joinerOn) {
                    joinerOn = false;
                } else if (!checkSubchar(charDec)) {
                    tryNext = false;
                    break;
                }
            }
            charDecList.push(charDec);
            
            pos = pos.translate(0, charDec < 65536 ? 1 : 2);
        } while (tryNext);

        if (charDecList.length == 0) {
            this._statusBarItem.hide();
            return;
        }

        // Update the status bar
        let msg = '';
        let unis = [];
        for (var i in charDecList) {
            let hex = toHex(charDecList[i], 4);
            msg += ' U+' + hex;
            unis.push(hex);
        }

        this._characters = editor.document.getText(new Range(cursorPos, pos));;
        this._unicodeDecList = charDecList;
        this._unicodeHexList = unis;

        this._statusBarItem.text = `$(file-binary) ${msg}`;
        this._statusBarItem.command = 'cursorCharcode.detailUnicodeInfo';
        this._statusBarItem.show();
    }

    dispose() {
        this._statusBarItem.dispose();
    }
}

class CharCodeController {
    private _display: CharCodeDisplay;
    private _disposable: Disposable;

    constructor(display: CharCodeDisplay) {
        this._display = display;

        // subscribe to selection change and editor activation events
        let subscriptions: Disposable[] = [];
        window.onDidChangeTextEditorSelection(this._onEvent, this, subscriptions);
        window.onDidChangeActiveTextEditor(this._onEvent, this, subscriptions);

        this._display.updateCharacterCode();
        this._disposable = Disposable.from(...subscriptions);
    }

    dispose() {
        this._disposable.dispose();
    }

    private _onEvent() {
        this._display.updateCharacterCode();
    }
}